pkg/image/importer/importer.go

```diff
@@ -0,0 +1,763 @@
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/fielderrors"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/dockerregistry"
+	"github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/api/dockerpre012"
+)
+
+// Add a dockerregistry.Client to the passed context with this key to support v1 Docker registry importing
+const ContextKeyV1RegistryClient = "v1-registry-client"
+
+// Interface loads images into an image stream import request.
+type Interface interface {
+	Import(cxt *AuthContext, isi *api.ImageStreamImport) error
+}
+
+// ErrNotV2Registry is returned when the server does not report itself as a V2 Docker registry
+type ErrNotV2Registry struct {
+	Registry string
+}
+
+func (e *ErrNotV2Registry) Error() string {
+	return fmt.Sprintf("endpoint %q does not support v2 API", e.Registry)
+}
+
+// ImageStreamImport implements an import strategy for Docker images. It keeps a cache of images
+// per distinct auth context to reduce duplicate loads
+type ImageStreamImporter struct {
+	maximumTagsPerRepo int
+
+	limiter util.RateLimiter
+
+	imageCache map[*AuthContext]map[location]*api.Image
+}
+
+// NewImageStreamImport creates an importer that will load images from a remote Docker registry into an
+// ImageStreamImport object. Limiter may be nil. This implementation is not thread safe.
+func NewImageStreamImporter(maximumTagsPerRepo int, limiter util.RateLimiter) *ImageStreamImporter {
+	if limiter == nil {
+		limiter = util.NewFakeRateLimiter()
+	}
+	return &ImageStreamImporter{
+		maximumTagsPerRepo: maximumTagsPerRepo,
+
+		limiter: limiter,
+
+		imageCache: make(map[*AuthContext]map[location]*api.Image),
+	}
+}
+
+// contextImageCache returns the image cache entry for a context.
+func (i *ImageStreamImporter) contextImageCache(ctx *AuthContext) map[location]*api.Image {
+	cache := i.imageCache[ctx]
+	if cache == nil {
+		cache = make(map[location]*api.Image)
+		i.imageCache[ctx] = cache
+	}
+	return cache
+}
+
+// Import tries to complete the provided isi object with images loaded from remote registries.
+func (i *ImageStreamImporter) Import(ctx *AuthContext, isi *api.ImageStreamImport) error {
+	cache := i.contextImageCache(ctx)
+	importImages(ctx, isi, cache, i.limiter)
+	importFromRepository(ctx, isi, i.maximumTagsPerRepo, cache, i.limiter)
+	return nil
+}
+
+// importImages updates the passed ImageStreamImport object and sets Status for each image based on whether the import
+// succeeded or failed. Cache is updated with any loaded images. Limiter is optional and controls how fast images are updated.
+func importImages(ctx *AuthContext, isi *api.ImageStreamImport, cache map[location]*api.Image, limiter util.RateLimiter) {
+	tags := make(map[location][]int)
+	ids := make(map[location][]int)
```

Comment:
`// maps from (registryURL,repoName) to list of isi.Spec.Images indexes that used it.  This is used while building status to relate repository information to associated Images`

That took a while.